### PR TITLE
Rover:Publish nav data to OSD

### PR DESCRIPTION
**Pull request by Hwurzburg**
Allows OSD display of WP#,bearing/distance and crosstrack error in all NAV modes(GUIDED,AUTO,RTL,SMARTRTL)....like other vehicles..

crosstrack in LOITER  (a boat only mode) displays distance to destination which is nice for tuning LOITER params...

when not in a nav mode, the current unexecuted WP info is displayed....ie bearing,distance,etc to next waypoint if AUTO is re-enabled...

### CODE REVIEW

File: APMrover2/Rover.cpp [hunk 1]
```diff
@@ -324,6 +327,33 @@ void Rover::update_mission(void)
     }
 }
 
+#if OSD_ENABLED == ENABLED
+void Rover::publish_osd_info()
+{
+    AP_OSD::NavInfo nav_info;
+    bool navmode = (control_mode == &mode_auto) || (control_mode == &mode_rtl) || (control_mode == &mode_smartrtl) || (control_mode == &mode_guided);
```
Comment (peterbarker) [inline APMrover2/Rover.cpp]: Explicitly calling modes out isn't a great thing to do.  Using methods on the current flight mode is the appropriate thing to do.

https://github.com/ardupilot/ardupilot/blob/master/APMrover2/GCS_Mavlink.cpp#L100 can be used as a reference for this.

**Comment by rmackay9** [comment]
This is close to what should be in the publish_osd_info function:

>     AP_OSD::NavInfo nav_info = {};
> 
>     // special handling for cross track error in loiter
>     if (control_mode == &mode_loiter) {
>         nav_info.wp_xtrack_error = control_mode->get_distance_to_destination();
>     } else {
>         nav_info.wp_xtrack_error = control_mode->crosstrack_error();
>     }
>     nav_info.wp_distance = control_mode->get_distance_to_destination();
>     nav_info.wp_bearing = control_mode->wp_bearing();
>     if (control_mode->is_autopilot_mode()) {
>         nav_info.wp_number = mode_auto.mission.get_current_nav_index();
>     }
>     osd.set_nav_info(nav_info);
> 

The only messy bit is still the mission item number.  We should really add a new accessor called get_wp_number() to the mode class (similar in some ways to the wp_bearing() method) and then return 0 for loiter.  For all other modes it can probably use the "if (control_mode->is_autopilot_mode()) {" stuff above.

File: APMrover2/Rover.cpp (hunk 1 above)
Comment (Hwurzburg) [reply APMrover2/Rover.cpp]: fixed per Randy's suggestions

**Comment by Hwurzburg** [comment]
fixed...working version is close to what you suggested....
SITL tested.....apparently LOITER reports as an autopilot mode so had to be explicitly excluded...

**Comment by rmackay9** [comment]
Merged with a couple of minor changes.. txs!

**Pull request closed without merging**